### ISSUE TITLE
type gets widened on declaration emit when inferred in destructuring context

### ISSUE THREAD

**Issue by h-joo** [microsoft/TypeScript#55439]
### 🔎 Search Terms

declaration, widened, d.ts, destructuring

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries

### ⏯ Playground Link

https://www.typescriptlang.org/play?jsx=0&module=1&ts=5.1.6#code/GYVwdgxgLglg9mABMOcAUBKAXIg3gTxwEYBfPAKESsQCcBTKEGpXRQxIxE878ugDwAOcGlEQQEAZzGt8iALyIADFwXJUmANxA

### 💻 Code

```ts
function foo(): {y: 1} {
    return { y: 1 }
}

export const { y = 0 } = foo();
```

### 🙁 Actual behavior

`y` is `number` in `d.ts` files, although the language service says `y` is `1|0`

### 🙂 Expected behavior

`y` is `1|0` in `d.ts` files.

### Additional information about the issue

Without the `= 0` assignment `y` properly becomes type `1` in d.ts

**Comment by MartinJohns** [comment]
Workaround: `0 as const`.

**Comment by dragomirtitian** [comment]
@Andarist I found a very similar issue. In the code below `x` is typed in ts as `0 | 1`, but is typed as `number` in declarations.
```ts
export const x = Math.random() ? 1 : 0 // 0 | 1
```
[Playground](https://www.typescriptlang.org/play?ts=5.2.2#code/KYDwDg9gTgLgBAYwgOwM7xHAvHAsgQxgAsA6KfZAEwgFsAKASjgH44BGOALjgAYBYAFAB6IXDHiJcAHrNBggJYAzOnExZ1cAExMA3oImCAvkA)

Will your fix  work for this example as well ?

**Comment by Andarist** [comment]
Yes, I think so - what Jake commented on here is basically the same as your case above: https://github.com/microsoft/TypeScript/pull/55445/files#r1300377161

**Comment by dragomirtitian** [comment]
@Andarist Oh. Yeah. Forgot about that. Thanks 😅
